blockchain/consensus/narwhal_cns/src/proposer.rs

```diff
@@ -0,0 +1,420 @@
+
+        // The following is based on the `NodeRestarter` in the narwhal repo.
+
+        // Get a store for the epoch of the committee (although currently there won't be reconfiguration.)
+        let mut store_path = self.storage_base_path.clone();
+        store_path.push(format!("epoch{}", self.committee.epoch()));
+        let store = NodeStorage::reopen(store_path);
+
+        let registry = prometheus::Registry::default();
+        let name = self.keypair.public().clone();
+
+        // Lazily connect to the worker which isn't running yet.
+        let transactions_client = make_transactions_client(&name, &self.committee)?;
+
+        // Start the Primary and a Worker. They will ask the execution state where to resume from.
+        let mut handles = Vec::new();
+
+        let primary_handles = Node::spawn_primary(
+            self.keypair,
+            Arc::new(ArcSwap::new(Arc::new(self.committee.clone()))),
+            &store,
+            self.parameters.clone(),
+            true,
+            execution_state,
+            &registry,
+        )
+        .await?;
+
+        let worker_handles = Node::spawn_workers(
+            name,
+            vec![0],
+            Arc::new(ArcSwap::new(Arc::new(self.committee))),
+            &store,
+            self.parameters,
+            &registry,
+        );
+
+        handles.extend(tokio_to_async_std(primary_handles));
+        handles.extend(tokio_to_async_std(worker_handles));
+
+        // Subscribe to local chain extensions.
+        let head_change_rx = chain_store.publisher().subscribe();
+
+        // Get the current head of the chain.
+        let current_head = chain_store.heaviest_tipset().await;
+
+        // NOTE: Instead of passig in the current head, we could use `chain_store.sub_header_changes` and `chain_store.next_header_change`,
+        // which has extra machinery to replay the current head, but it involves extra bookkeeping, channels and tasks, and also sets a
+        // lower limit for lagging, although that in our case wouldn't matter since we create blocks at the pace we can consume them.
+        let state_manager_clone = state_manager.clone();
+        handles.push(task::spawn(async move {
+            if let Err(e) = handle_head_changes(
+                state_manager_clone,
+                submitter,
+                self.validator_addresses,
+                current_head,
+                head_change_rx,
+                head_tx,
+                output_rx,
+            )
+            .await
+            {
+                error!("Error handling head changes: {}", e)
+            }
+        }));
+
+        handles.push(task::spawn(async move {
+            if let Err(e) =
+                handle_head_watch(state_manager, mpool, head_rx, transactions_client).await
+            {
+                error!("Error handling head watch: {}", e)
+            }
+        }));
+
+        Ok(handles)
+    }
+}
+
+fn tokio_to_async_std(
```

Comment:
Out of sheer ignorance, why do you cast tokio handles into async_std ones?

Reply:
I promised our `daemon.rs` to return `async_std::task::JoinHandle` in `composition.rs`, but the ones from Narwhal are `tokio` ones. I thought it would be good to return the handles of these long lived processes so they can be canceled upon exit like the other tasks in `daemon.rs`. Not sure if cancel works, really, but fingers crossed.

Reply:
That make sense, I completely missed that the daemon was expecting an async_std. I think the cancel approach is a really good idea (we use something similar in Go to cancel all the routines involved in the process).